### PR TITLE
Azure Storage signature generator using SpanWriter

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
I rewrote the routine generating Azure Storage access signature to now use the SpanWriter (it was using our low level primitive APIs before). The code is now much simpler and equally perfromant. 

Generating Azure Storage access signaturess using SpanWriter does not allocate any GC heap memory; the currently prescribed (by MSDN) way using System.String APIs allocates 1KB per signature. Also, the SpanWriter implementation cuts the number of CPU instructions in half and elapsed time by 3x.

Here are perfromance results:
- StorageBench.Msdn : MSDN sample showing how to generate the header using .NET APIs
- StorageBench.Writer : corefxlab APIs using SpanWriter
- StorageBench.WriterUtf8: corefxlab APIs using SpanWriter and Utf8String
- StorageBench.Primitive: same behavior using low level primitives as opposed to SpanWriter

Test Name | Metric | AVERAGE
:----------------------- | :--------------------------------------------- | ---------:
StorageBench.Msdn | Duration | 0.009
StorageBench.Msdn | Instructions Retired | 40700
StorageBench.Msdn | Cache Misses | 61.44
StorageBench.Msdn | Allocation Size on Benchmark   Execution Thread | 1280
StorageBench.Primitive | Duration | 0.003
StorageBench.Primitive | Instructions Retired | 21500
StorageBench.Primitive | Cache Misses | 24.576
StorageBench.Primitive | Allocation Size on Benchmark   Execution Thread | 0
StorageBench.Writer | Duration | 0.003
StorageBench.Writer | Instructions Retired | 22100
StorageBench.Writer | Cache Misses | 4.096
StorageBench.Writer | Allocation Size on Benchmark   Execution Thread | 0
StorageBench.WriterUtf8 | Duration | 0.003
StorageBench.WriterUtf8 | Instructions Retired | 21100
StorageBench.WriterUtf8 | Cache Misses | 8.192
StorageBench.WriterUtf8 | Allocation Size on Benchmark   Execution Thread | 0

I moved the old code (using primitive formatters) to the benchmark dll so we can keep comparing perfromance as the writer design changes.

cc: @ahsonkhan, @GrabYourPitchforks, @joshfree, @JeffreyRichter

### CODE REVIEW

File: src/System.Azure.Experimental/System/Azure/StorageAccessSignature.cs [hunk 1]
```diff
@@ -6,79 +6,56 @@
...
 using System.Buffers.Text;
 using System.Buffers.Cryptography;
-using System.Text;
+using System.Text.Utf8;
 
 namespace System.Azure.Authentication
 {
     public static class StorageAccessSignature
     {
+        static Utf8String s_emptyHeaders = (Utf8String)"\n\n\n\n\n\n\n\n\n\n\nx-ms-date:"; // this wont be beeded once we have UTF8 literals
```
Comment (ahsonkhan) [inline src/System.Azure.Experimental/System/Azure/StorageAccessSignature.cs]: nit: typo beeded -> needed

File: src/System.Azure.Experimental/System/Azure/StorageAccessSignature.cs [hunk 2]
```diff
@@ -6,79 +6,56 @@
...
-                if (output.Length < 3)
-                {
-                    bytesWritten = 0;
-                    return false;
-                }
-                s_GET.CopyTo(output);
-                bytesWritten += s_GET.Length;
+                var writer = new SpanWriter(output);
+                writer.WriteLine(verb);
+                writer.Write("\n\n\n\n\n\n\n\n\n\n\nx-ms-date:");
```
Comment (ahsonkhan) [inline src/System.Azure.Experimental/System/Azure/StorageAccessSignature.cs]: Use `s_emptyHeaders`?

File: tests/Benchmarks/AzureBench.cs [hunk 3]
```diff
@@ -239,3 +238,175 @@ public static bool TryWritePrimitives(Span<byte> output, Sha256 hash, string key
     const int AuthenticationHeaderBufferSize = 256;
 }
 
+public class StorageBench
```
Comment (ahsonkhan) [inline tests/Benchmarks/AzureBench.cs]: nit: split into separate file

File: tests/Benchmarks/AzureBench.cs [hunk 4]
```diff
@@ -15,7 +15,7 @@
 using System.Text.Encodings.Web.Utf8;
 using System.Text.Utf8;
 
-public class AzureBench
+public class CosmosDbBench
```
Comment (ahsonkhan) [inline tests/Benchmarks/AzureBench.cs]: Why not move these to the Azure Test project and add to the perf harness so you get the granularity of running just these performance tests (similar to others like [System.Binary.Base64.Tests](https://github.com/dotnet/corefxlab/blob/master/tests/System.Binary.Base64.Tests/PerformanceTests.cs))?

File: src/System.Azure.Experimental/System/Azure/StorageAccessSignature.cs (hunk 1 above)
Comment (KrzysztofCwalina) [reply src/System.Azure.Experimental/System/Azure/StorageAccessSignature.cs]: fixed. thanks

File: src/System.Azure.Experimental/System/Azure/StorageAccessSignature.cs (hunk 2 above)
Comment (KrzysztofCwalina) [reply src/System.Azure.Experimental/System/Azure/StorageAccessSignature.cs]: s_emptyHeaders is Utf8String. I wanted this one to be UTF16

File: tests/Benchmarks/AzureBench.cs (hunk 3 above)
Comment (KrzysztofCwalina) [reply tests/Benchmarks/AzureBench.cs]: done

File: tests/Benchmarks/AzureBench.cs (hunk 4 above)
Comment (KrzysztofCwalina) [reply tests/Benchmarks/AzureBench.cs]: run.bat StorageBench runs just storage tests. i.e. good enough granularity.

**Pull request merged**